### ISSUE TITLE
Put Mapael through JSHint

### ISSUE THREAD

**Issue by Indigo744** [neveldo/jQuery-Mapael#89]
It would be good if Mapael passes the JSHint verification.

http://jshint.com

I suggest the following permanent options:
```
/* jshint jquery:true, unused:true */
```
- `jquery`: defines globals exposed by the jQuery JavaScript library
- `unused`: warns when a variable is defined and never used (note: this is set to `true` instead of `strict` to avoid warning about unused parameters in functions).

The following globals needs to be declared:
```
/* globals Raphael */
/* globals exports, module, require */ // CommonJS
/* globals define */ // AMD
```

Also, in a first time, I suggest adding the following options to reduce the number of warnings:
```
/* jshint laxcomma:true, laxbreak:true, expr:true */
```
- `laxcomma`: suppresses warnings about comma-first coding style
- `laxbreak`: suppresses most of the warnings about possibly unsafe line breakings in your code
- `expr`: suppresses warnings about the use of expressions where normally you would expect to see assignments or function calls

Number of warning remaining: **34 warnings**

There are a lot of `Use '===' to compare with ...`. 
I think Mapael should indeed use the *triple eq* more often.

**Comment by neveldo** [comment]
Hello,

Nice proposal, It will help to keep the code clean ! I will add a .jshintrc file in the project root (in order to also check maps files and examples) with your suggested configuration and make the code fixes regarding the 34 warnings.

**Comment by Indigo744** [comment]
Ah yes, maybe it's better to put the jshint options in a separate file!

If you want, I can start a PR in order to resolve some of the warning (the easiest ones?).

Also, I noticed that in a recent commit there are now a mix of tab and space characters in the code.
Do you prefer proper tab or 4-space tab? I usually use the later in my own code. Easier to fiddle with the alignment using spaces :-)

**Comment by neveldo** [comment]
Of course, feel free to start a PR for cleaning the code if you can :)

For the tab/space, I have no particular preference. I checked mapael dependencies, Raphael.js uses spaces, and jQuery uses tabs ... 

In the examples I added yesterday, I have used 4 spaces for indentation, so I think we can switch jquery.mapael.js (and maps) to 4-spaces indentation !

**Comment by Indigo744** [comment]
After these 2 PR, number of warning remaining: **23 warnings**

**Comment by neveldo** [comment]
.jshintrc file added : https://github.com/neveldo/jQuery-Mapael/commit/0fe9062a5f7dd31bb83f8b1c46b3ff23563074cf . Feel free to update the jshint options in the .jshintrc, I will check the errors as soon as possible.

**Comment by neveldo** [comment]
Thank you, I have merge your two PRs :)

By the way, I'm very glad you contribute to the project a lot since few days ! Could I add you (Indigo744) into a "thanks" section into the top comment of jquery.mapael.js ? 

In any case, I hope that Mapael fits with your needs for your project :)

**Comment by Indigo744** [comment]
Sorry for the spam :-) I prefer to track the changes incrementally in multiple PR in order to not break anything. Is it ok for you?

Mapael is perfect for my project. I started submitting PR with the features I needed for my project, but then I don't know why but I kind of fell in love with this little piece of code. So yeah I'm sticking around. And I'm learning Git/Github as well so it's good.
Feel free to add something if you want. But don't consider it mandatory, it's just my way to say "*thank you for this plugin*" ;-)

**Comment by neveldo** [comment]
It's ok for me, I prefer to have lightweight pull requests to review and to merge.

I'm glad Mapael fits with your project needs, and that you fell in love with it ;-) . So I will not miss to add a thank section, in order to say "thank you for all your contributions" !

**Comment by Indigo744** [comment]
Why is there a `"nonbsp": true` in the .jshintrc file ?
I fail to see the need for such an option.

**Comment by neveldo** [comment]
Hello,

> This option warns about "non-breaking whitespace" characters. These characters can be entered with > option-space on Mac computers and have a potential of breaking non-UTF8 web pages.

I have added this options to prevent from adding non-breaking whitespace in the source code, I think there is no reason to have this character within the code. I suspect the weird characters fixed here : https://github.com/neveldo/jQuery-Mapael/commit/f855c2113a87c7dc1d877e0db725d500997f0a33 to be non-breaking whitespace. Am I wrong ?

**Comment by Indigo744** [comment]
Oh right, I thought this option *disabled* the checking of non-breaking whitespace XD
My bad!

**Comment by Indigo744** [comment]
Now after that, there seems to be almost no warning (except a silly one about `mousewheels`, but anyway).

I wonder if we remove the `expr: true` options. This option may hide some real issue. 

Without this option, we get around 30 warnings `Expected an assignment or function call and instead saw an expression`.
They are all related to a one-liner pattern that you use a lot:
```myvar && do_something_with(my_var);```
This pattern could be rewrite to:
```if (myvar) do_something_with(my_var);```

But I don't want to force any "style" of coding on you. If it's a pattern that you like, then ok :-)

**Comment by neveldo** [comment]
Hello,

Thanks for all the JSHint fixes you have done on Mapael. Don't worry, I'm not stuck to a particular coding style. Actually, I agree with you, the if() {} syntax should be even more clear than the current syntax I think.

I wonder how to deal with the warning about the "mousewheels" argument of the factory function. Does JSHint allow to set some kind of exception for a specific function ?

**Comment by Indigo744** [comment]
We could use the following code:
```
}(function ($, Raphael, mousewheel) { // jshint ignore:line
```
Or we could just ignore this warning.

**Comment by Indigo744** [comment]
2 new PRs :-)

**Comment by neveldo** [comment]
Thank you, I will review them as soon as possible. Regarding the JSHint warning, I thin we could just ignore it as you suggested :)

**Comment by Indigo744** [comment]
About that remaining warning, it seems that when doing a `grunt`, the task fails and the process stops because of it. Only a `grunt --force` make it able to continue to minification.
So I'm still wondering if we should do something about it?

**Comment by neveldo** [comment]
You are right, I have noticed the warning when running grunt on mapael. I suggest to use the `// jshint ignore:line` comment in order to be able to run grunt sucessfully in normal mode, but maybe you have a better proposal ?

**Comment by Indigo744** [comment]
Sadly I don't have a better proposal. I don't really like the idea of impacting the source code with jshint related comments, but it doesn't seems to be another way :/

**Comment by neveldo** [comment]
Done : https://github.com/neveldo/jQuery-Mapael/commit/f60e066c01a93874f2bcc6c3b52e14ebc476647b :)
